solver/src/settlement.rs

```diff
@@ -111,26 +117,53 @@ impl Settlement {
         Ok(cursor.into_inner())
     }
 
-    // For now this computes the total surplus of all EOA trades.
-    pub fn objective_value(&self) -> U256 {
-        // TODO: in order to achieve a fair comparison between different solutions, we need to normalize the price
-        // vector as otherwise any price vector scaling by factor x would also scale the objective value by that factor
-        // without affecting the validity of the solution.
-        match self.trades.iter().fold(Some(U256::zero()), |acc, trade| {
-            let sell_token_price = self
+    fn total_surplus(
+        &self,
+        normalizing_prices: &HashMap<H160, BigRational>,
+    ) -> Option<BigRational> {
+        self.trades.iter().fold(Some(num::zero()), |acc, trade| {
+            let sell_token_clearing_price = self
                 .clearing_prices
                 .get(&trade.order.sell_token)
-                .expect("Solution with trade but without price for sell token");
-            let buy_token_price = self
+                .expect("Solution with trade but without price for sell token")
+                .to_big_rational();
+            let buy_token_clearing_price = self
                 .clearing_prices
                 .get(&trade.order.buy_token)
+                .expect("Solution with trade but without price for buy token")
+                .to_big_rational();
+
+            let sell_token_external_price = normalizing_prices
+                .get(&trade.order.sell_token)
                 .expect("Solution with trade but without price for sell token");
-            acc?.checked_add(trade.surplus(*sell_token_price, *buy_token_price)?)
-        }) {
+            let buy_token_external_price = normalizing_prices
+                .get(&trade.order.buy_token)
+                .expect("Solution with trade but without price for buy token");
+
+            let surplus = &trade.surplus(&sell_token_clearing_price, &buy_token_clearing_price)?;
+            let normalized_surplus = match trade.order.kind {
+                OrderKind::Sell => surplus * buy_token_external_price / buy_token_clearing_price,
+                OrderKind::Buy => surplus * sell_token_external_price / sell_token_clearing_price,
+            };
+            acc?.checked_add(&normalized_surplus)
+        })
+    }
+
+    // Objective is re-computed using external prices.
+    fn objective_value_recomputed(
+        &self,
+        external_prices: &HashMap<H160, BigRational>,
+    ) -> Option<BigRational> {
+        self.total_surplus(external_prices)
+    }
+
+    // For now this computes the total surplus of all EOA trades.
+    pub fn objective_value(&self, external_prices: &HashMap<H160, BigRational>) -> BigRational {
```

Comment:
I see that this method is now just encapsulating its recomputed counterpart with an additional overflow handling, do we really need this or could it all fit into a single place?

Reply:
Yes, fixed in 779d70 . Thanks!